In [1]:
# remove hash to install package
# !pip3 install owslib==0.25.0 fiona==1.8.21 geopandas==0.10.2 requests==2.28.0 folium==0.12.1

In [1]:
from owslib.wfs import WebFeatureService
import geopandas
import folium
import io
import zipfile
import pandas as pd
import os
from urllib.request import urlretrieve

## Download external data from AURIN

In [2]:
# User credential to connect with API
WFS_USERNAME = 'ilcag'
WFS_PASSWORD= 'NCpx3cztUjshvRDb'
WFS_URL='https://adp.aurin.org.au/geoserver/wfs'

In [4]:
# Connect with API
adp_client = WebFeatureService(url=WFS_URL,username=WFS_USERNAME, password=WFS_PASSWORD, version='2.0.0')

In [5]:
def download_aurin_df(type_name, file_name):
    """
        This function downloads the dataset from AURIN API
        
        type_name: dataset identifier from the website
        file_name: output file name 
    """

    output_dir = '../data/abs'
    
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)
        
    response = adp_client.getfeature(typename=type_name)
    out = open(f'{output_dir}/{file_name}.gml', 'wb')
    out.write(response.read())
    out.close()
    return geopandas.read_file(f'{output_dir}/{file_name}.gml')



In [8]:
# Download selected external data

pop_df = download_aurin_df('datasource-AU_Govt_ABS-UoM_AURIN_DB_3:abs_regional_population_sa2_2001_2021', 
                           '2021_population_census')

In [11]:
sa2_bound = download_aurin_df('datasource-AU_Govt_ABS-UoM_AURIN_DB_GeoLevel:sa2_2016_aust',
                        'sa2_boundaries')

In [14]:
poa_bound = download_aurin_df('datasource-AU_Govt_ABS-UoM_AURIN_DB_GeoLevel:poa_2016_aust',
                        'poa_boundaries')

In [15]:
sa2_income = download_aurin_df('datasource-AU_Govt_ABS-UoM_AURIN_DB_3:abs_personal_income_total_income_distribution_sa2_2017_18', 
                        'sa2_income')

In [16]:
# Select 2021 population census

area_id = ['gml_id', 'primaryindex', 'state_code_2016', 'sa2_maincode_2016',
           'sa2_name_2016']
col_2021 = [x for x in pop_df.columns if '2021' in x or '2020_21' in x]
pop_21 = pop_df[area_id + col_2021]

In [17]:
pop_21.shape

(2292, 18)

There should be 2,310 SA2 regions.

In [18]:
# ERP refers to estimated resident population
pop_21.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2292 entries, 0 to 2291
Data columns (total 18 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   gml_id                           2292 non-null   object 
 1   primaryindex                     2292 non-null   int64  
 2   state_code_2016                  2292 non-null   int64  
 3   sa2_maincode_2016                2292 non-null   int64  
 4   sa2_name_2016                    2292 non-null   object 
 5   erp_2021                         2292 non-null   int64  
 6   erp_change_number_2020_21        2292 non-null   int64  
 7   erp_change_per_cent_2020_21      2292 non-null   float64
 8   pop_density_2021_people_per_km2  2292 non-null   float64
 9   births_2020_21                   2288 non-null   float64
 10  deaths_2020_21                   2288 non-null   float64
 11  natural_increase_2020_21         2288 non-null   float64
 12  internal_arrivals_20

Do we want to include these regions in the data???
If so, will we include the attributes with missing data?
If yes, how?
If no, why?

In [19]:
pop_21[pop_21["births_2020_21"].isnull()]

gml_id  primaryindex  \
2231  abs_regional_population_sa2_2001_2021.2292          2292   
2289  abs_regional_population_sa2_2001_2021.2289          2289   
2290  abs_regional_population_sa2_2001_2021.2290          2290   
2291  abs_regional_population_sa2_2001_2021.2291          2291   

      state_code_2016  sa2_maincode_2016            sa2_name_2016  erp_2021  \
2231                9          901041004           Norfolk Island      1749   
2289                9          901011001         Christmas Island      1979   
2290                9          901021002  Cocos (Keeling) Islands       579   
2291                9          901031003               Jervis Bay       397   

      erp_change_number_2020_21  erp_change_per_cent_2020_21  \
2231                         14                       0.8069   
2289                         15                       0.7637   
2290                          6                       1.0471   
2291                          2                       0.5063   

      pop_density_2021_people_per_km2  births_2020_21  deaths_2020_21  \
2231                        45.251202             NaN             NaN   
2289                        14.537000             NaN             NaN   
2290                        42.212601             NaN             NaN   
2291                         5.854300             NaN             NaN   

      natural_increase_2020_21  internal_arrivals_2020_21  \
2231                       NaN                        NaN   
2289                       NaN                        NaN   
2290                       NaN                        NaN   
2291                       NaN                        NaN   

      internal_departures_2020_21  net_internal_migration_2020_21  \
2231                          NaN                             NaN   
2289                          NaN                             NaN   
2290                          NaN                             NaN   
2291                          NaN                             NaN   

      overseas_arrivals_2020_21  overseas_departures_2020_21  \
2231                        NaN                          NaN   
2289                        NaN                          NaN   
2290                        NaN                          NaN   
2291                        NaN                          NaN   

      net_overseas_migration_2020_21  
2231                             NaN  
2289                             NaN  
2290                             NaN  
2291                             NaN

---
## Download Postcode to SA2 table

In [20]:
def download_url(url, filename):
    '''
        This function downloads data from the specified url.

        url: url of specified website
        filename: output file name
    '''
    output_dir = '../data/abs'
    
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)
        
    print(f"Begin downloading {filename} data")
    output_dir = f"{output_dir}/{filename}.zip"
    urlretrieve(url, output_dir)
    print(f"Completed")

In [21]:
download_url('https://www.abs.gov.au/AUSSTATS/subscriber.nsf/log?openagent&1270055006_CG_POSTCODE_2011_SA2_2011.zip&1270.0.55.006&Data%20Cubes&70A3CE8A2E6F9A6BCA257A29001979B2&0&July%202011&27.06.2012&Latest'
             , 'poa_sa2_lookup')

Begin downloading poa_sa2_lookup data
Completed


In [22]:
# Open zipfile
unzip_poa_sa2 = zipfile.ZipFile('../data/abs/poa_sa2_lookup.zip') 

In [23]:
poa_to_sa2 = pd.read_excel(unzip_poa_sa2.open('1270055006_CG_POSTCODE_2011_SA2_2011.xls')
                        , sheet_name='Table 3', skiprows=5)

In [24]:
poa_to_sa2 = poa_to_sa2.dropna()

There are 2162 SA2 codes in Postcode to SA2 data, but there are 2292. Try to find newer data for poa_to_sa2?

In [25]:
poa_to_sa2["SA2_MAINCODE_2011"].unique().size

2162

In [26]:
poa_bound

gml_id  primaryindex  objectid  poa_code_2016 poa_name_2016  \
0        poa_2016_aust.1             1         1            800          0800   
1        poa_2016_aust.2             2         2            810          0810   
2        poa_2016_aust.3             3         3            812          0812   
3        poa_2016_aust.4             4         4            815          0815   
4        poa_2016_aust.5             5         5            820          0820   
...                  ...           ...       ...            ...           ...   
2665  poa_2016_aust.2639          2639      2639           7268          7268   
2666  poa_2016_aust.2640          2640      2640           7270          7270   
2667  poa_2016_aust.2641          2641      2641           7275          7275   
2668  poa_2016_aust.2642          2642      2642           7276          7276   
2669  poa_2016_aust.2643          2643      2643           7277          7277   

      area_albers_sqkm                                           geometry  
0               3.1734  POLYGON ((130.83450 -12.45800, 130.83390 -12.4...  
1              23.7902  POLYGON ((130.84710 -12.37750, 130.84730 -12.3...  
2              35.8899  POLYGON ((130.89190 -12.36880, 130.89220 -12.3...  
3               0.6381  POLYGON ((130.87240 -12.37650, 130.87230 -12.3...  
4              39.0462  POLYGON ((130.83500 -12.43010, 130.83510 -12.4...  
...                ...                                                ...  
2665          162.4060  POLYGON ((147.20120 -41.25830, 147.20130 -41.2...  
2666          265.1620  MULTIPOLYGON (((146.82300 -41.13960, 146.82290...  
2667          340.3100  POLYGON ((146.84830 -41.24130, 146.84850 -41.2...  
2668            5.9926  POLYGON ((146.97820 -41.28340, 146.97820 -41.2...  
2669           98.6653  MULTIPOLYGON (((147.08800 -41.38310, 147.08800...  

[2670 rows x 7 columns]

In [27]:
sa2_bound

gml_id  primaryindex  sa2_maincode_2016  \
0        sa2_2016_aust.1             1          101021007   
1        sa2_2016_aust.2             2          101021008   
2        sa2_2016_aust.9             9          101031015   
3       sa2_2016_aust.10            10          101031016   
4       sa2_2016_aust.11            11          101041017   
...                  ...           ...                ...   
2305  sa2_2016_aust.2305          2305          901011001   
2306  sa2_2016_aust.2306          2306          901021002   
2307  sa2_2016_aust.2307          2307          901031003   
2308  sa2_2016_aust.2308          2308          901041004   
2309  sa2_2016_aust.2309          2309          997979799   

      sa2_5digitcode_2016                         sa2_name_2016  \
0                   11007                             Braidwood   
1                   11008                               Karabar   
2                   11015                          Cooma Region   
3                   11016                 Jindabyne - Berridale   
4                   11017                          Batemans Bay   
...                   ...                                   ...   
2305                91001                      Christmas Island   
2306                91002               Cocos (Keeling) Islands   
2307                91003                            Jervis Bay   
2308                91004                        Norfolk Island   
2309                99799  Migratory - Offshore - Shipping (OT)   

      sa3_code_2016                         sa3_name_2016  sa4_code_2016  \
0             10102                            Queanbeyan            101   
1             10102                            Queanbeyan            101   
2             10103                       Snowy Mountains            101   
3             10103                       Snowy Mountains            101   
4             10104                           South Coast            101   
...             ...                                   ...            ...   
2305          90101                      Christmas Island            901   
2306          90102               Cocos (Keeling) Islands            901   
2307          90103                            Jervis Bay            901   
2308          90104                        Norfolk Island            901   
2309          99797  Migratory - Offshore - Shipping (OT)            997   

                             sa4_name_2016 gccsa_code_2016  \
0                           Capital Region           1RNSW   
1                           Capital Region           1RNSW   
2                           Capital Region           1RNSW   
3                           Capital Region           1RNSW   
4                           Capital Region           1RNSW   
...                                    ...             ...   
2305                     Other Territories           9OTER   
2306                     Other Territories           9OTER   
2307                     Other Territories           9OTER   
2308                     Other Territories           9OTER   
2309  Migratory - Offshore - Shipping (OT)           99799   

                           gccsa_name_2016  state_code_2016  \
0                              Rest of NSW                1   
1                              Rest of NSW                1   
2                              Rest of NSW                1   
3                              Rest of NSW                1   
4                              Rest of NSW                1   
...                                    ...              ...   
2305                     Other Territories                9   
2306                     Other Territories                9   
2307                     Other Territories                9   
2308                     Other Territories                9   
2309  Migratory - Offshore - Shipping (OT)                9   

        state_name_2016  area_albers_sqkm  \
0       New South Wa

In [28]:
poa_bound["geometry"] = poa_bound["geometry"].to_crs(epsg=4326)
sa2_bound["geometry"] = sa2_bound["geometry"].to_crs(epsg=4326)

In [29]:
poa_bound["centroid"] = poa_bound["geometry"].centroid

/var/folders/kn/0h56wsl91tz9kbw4pcdll7zw0000gn/T/ipykernel_2010/1735245261.py:1: UserWarning: Geometry is in a geographic CRS. Results from 'centroid' are likely incorrect. Use 'GeoSeries.to_crs()' to re-project geometries to a projected CRS before this operation.

  poa_bound["centroid"] = poa_bound["geometry"].centroid


In [30]:
poa_bound = poa_bound[poa_bound["centroid"].notnull()]

In [31]:
sa2_bound = sa2_bound[sa2_bound["geometry"].notnull()]

In [32]:
from shapely.validation import make_valid
sa2_bound["geometry"] = sa2_bound["geometry"].apply(make_valid)

/Users/patrick/opt/anaconda3/lib/python3.9/site-packages/geopandas/geodataframe.py:1351: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  super().__setitem__(key, value)


In [45]:
from shapely.geometry import Point
poa_list = poa_bound["centroid"].to_list()
sa2_list = []
for point in poa_list:
    contains = sa2_bound["geometry"].contains(Point(point))
    sa2_list.append(contains.index[contains])

In [34]:
sa2_list

[Int64Index([2130], dtype='int64'),
 Int64Index([1401], dtype='int64'),
 Int64Index([1454], dtype='int64'),
 Int64Index([1401], dtype='int64'),
 Int64Index([2139], dtype='int64'),
 Int64Index([2184], dtype='int64'),
 Int64Index([2117], dtype='int64'),
 Int64Index([2184], dtype='int64'),
 Int64Index([103], dtype='int64'),
 Int64Index([185], dtype='int64'),
 Int64Index([2308], dtype='int64'),
 Int64Index([2259], dtype='int64'),
 Int64Index([2269], dtype='int64'),
 Int64Index([1375], dtype='int64'),
 Int64Index([1943], dtype='int64'),
 Int64Index([1947], dtype='int64'),
 Int64Index([2165], dtype='int64'),
 Int64Index([2120], dtype='int64'),
 Int64Index([2117], dtype='int64'),
 Int64Index([1943], dtype='int64'),
 Int64Index([1946], dtype='int64'),
 Int64Index([2184], dtype='int64'),
 Int64Index([2120], dtype='int64'),
 Int64Index([2120], dtype='int64'),
 Int64Index([2184], dtype='int64'),
 Int64Index([2185], dtype='int64'),
 Int64Index([2176], dtype='int64'),
 Int64Index([2175], dtype='int

In [35]:
poa_bound["sa2_area"] = pd.Series(sa2_list)

In [74]:
sa2_bound

gml_id  primaryindex  sa2_maincode_2016  \
0        sa2_2016_aust.1             1          101021007   
1        sa2_2016_aust.2             2          101021008   
2        sa2_2016_aust.9             9          101031015   
3       sa2_2016_aust.10            10          101031016   
4       sa2_2016_aust.11            11          101041017   
...                  ...           ...                ...   
2302  sa2_2016_aust.2302          2302          801111141   
2305  sa2_2016_aust.2305          2305          901011001   
2306  sa2_2016_aust.2306          2306          901021002   
2307  sa2_2016_aust.2307          2307          901031003   
2308  sa2_2016_aust.2308          2308          901041004   

      sa2_5digitcode_2016            sa2_name_2016  sa3_code_2016  \
0                   11007                Braidwood          10102   
1                   11008                  Karabar          10102   
2                   11015             Cooma Region          10103   
3                   11016    Jindabyne - Berridale          10103   
4                   11017             Batemans Bay          10104   
...                   ...                      ...            ...   
2302                81141                  Namadgi          80111   
2305                91001         Christmas Island          90101   
2306                91002  Cocos (Keeling) Islands          90102   
2307                91003               Jervis Bay          90103   
2308                91004           Norfolk Island          90104   

                sa3_name_2016  sa4_code_2016                 sa4_name_2016  \
0                  Queanbeyan            101                Capital Region   
1                  Queanbeyan            101                Capital Region   
2             Snowy Mountains            101                Capital Region   
3             Snowy Mountains            101                Capital Region   
4                 South Coast            101                Capital Region   
...                       ...            ...                           ...   
2302       Urriarra - Namadgi            801  Australian Capital Territory   
2305         Christmas Island            901             Other Territories   
2306  Cocos (Keeling) Islands            901             Other Territories   
2307               Jervis Bay            901             Other Territories   
2308           Norfolk Island            901             Other Territories   

     gccsa_code_2016               gccsa_name_2016  state_code_2016  \
0              1RNSW                   Rest of NSW                1   
1              1RNSW                   Rest of NSW                1   
2              1RNSW                   Rest of NSW                1   
3              1RNSW                   Rest of NSW                1   
4              1RNSW                   Rest of NSW                1   
...              ...                           ...              ...   
2302           8ACTE  Australian Capital Territory                8   
2305           9OTER             Other Territories                9   
2306           9OTER             Other Territories                9   
2307           9OTER             Other Territories                9   
2308           9OTER             Other Territories                9   

                   state_name_2016  area_albers_sqkm  \
0                  New South Wales         3418.3525   
1                  New South Wales            6.9825   
2                  New South Wales         6250.8748   
3                  New South Wales         3939.5484   
4                  New South Wales           63.7074   
...                            ...               ...   
2302  Australian Capital Territory         1202.8930   
2305             Other Territories          136.1356   
2306             Other Territories           13.7163   
2307             Other Territories           67.8134   
2308             Other Territories           38.65

In [76]:
poa_w_sa2 = poa_bound.sjoin(sa2_bound[['sa2_maincode_2016', 'sa2_name_2016','geometry']], lsuffix="sa2_area", rsuffix=sa2_bound.index)

In [78]:
poa_w_sa2.head(20)

gml_id  primaryindex  objectid  poa_code_2016 poa_name_2016  \
0   poa_2016_aust.1             1         1            800          0800   
4   poa_2016_aust.5             5         5            820          0820   
0   poa_2016_aust.1             1         1            800          0800   
4   poa_2016_aust.5             5         5            820          0820   
0   poa_2016_aust.1             1         1            800          0800   
4   poa_2016_aust.5             5         5            820          0820   
0   poa_2016_aust.1             1         1            800          0800   
4   poa_2016_aust.5             5         5            820          0820   
1   poa_2016_aust.2             2         2            810          0810   
4   poa_2016_aust.5             5         5            820          0820   
1   poa_2016_aust.2             2         2            810          0810   
2   poa_2016_aust.3             3         3            812          0812   
4   poa_2016_aust.5             5         5            820          0820   
13  poa_2016_aust.7             7         7            828          0828   
1   poa_2016_aust.2             2         2            810          0810   
2   poa_2016_aust.3             3         3            812          0812   
4   poa_2016_aust.5             5         5            820          0820   
13  poa_2016_aust.7             7         7            828          0828   
1   poa_2016_aust.2             2         2            810          0810   
4   poa_2016_aust.5             5         5            820          0820   

    area_albers_sqkm                                           geometry  \
0             3.1734  POLYGON ((130.83451 -12.45799, 130.83391 -12.4...   
4            39.0462  POLYGON ((130.83501 -12.43009, 130.83511 -12.4...   
0             3.1734  POLYGON ((130.83451 -12.45799, 130.83391 -12.4...   
4            39.0462  POLYGON ((130.83501 -12.43009, 130.83511 -12.4...   
0             3.1734  POLYGON ((130.83451 -12.45799, 130.83391 -12.4...   
4            39.0462  POLYGON ((130.83501 -12.43009, 130.83511 -12.4...   
0             3.1734  POLYGON ((130.83451 -12.45799, 130.83391 -12.4...   
4            39.0462  POLYGON ((130.83501 -12.43009, 130.83511 -12.4...   
1            23.7902  POLYGON ((130.84711 -12.37749, 130.84731 -12.3...   
4            39.0462  POLYGON ((130.83501 -12.43009, 130.83511 -12.4...   
1            23.7902  POLYGON ((130.84711 -12.37749, 130.84731 -12.3...   
2            35.8899  POLYGON ((130.89191 -12.36879, 130.89221 -12.3...   
4            39.0462  POLYGON ((130.83501 -12.43009, 130.83511 -12.4...   
13           28.6532  POLYGON ((130.91771 -12.43139, 130.91401 -12.4...   
1            23.7902  POLYGON ((130.84711 -12.37749, 130.84731 -12.3...   
2            35.8899  POLYGON ((130.89191 -12.36879, 130.89221 -12.3...   
4            39.0462  POLYGON ((130.83501 -12.43009, 130.83511 -12.4...   
13           28.6532  POLYGON ((130.91771 -12.43139, 130.91401 -12.4...   
1            23.7902  POLYGON ((130.84711 -12.37749, 130.84731 -12.3...   
4            39.0462  POLYGON ((130.83501 -12.43009, 130.83511 -12.4...   

                       centroid                           sa2_area  \
0   POINT (130.84373 -12.46185)  Int64Index([2130], dtype='int64')   
4   POINT (130.86497 -12.42338)  Int64Index([2139], dtype='int64')   
0   POINT (130.84373 -12.46185)  Int64Index([2130], dtype='int64')   
4   POINT (130.86497 -12.42338)  Int64Index([2139], dtype='int64')   
0   POINT (130.84373 -12.46185)  Int64Index([2130], dtype='int64')   
4   POINT (130.86497 -12.42338)  Int64Index([2139], dtype='int64')   
0   POINT (130.84373 -12.46185)  Int64Index([2130], dtype='int64')   
4   POINT (130.86497 -12.42338)  Int64Index([2139], dtype='int64')   
1   POINT (130.87685 -12.36972)  Int64Index([1401], dtype='int64')   
4   POINT (130.86497 -12.42338)  Int64Index([2139], dtype='int64')   
1   POINT (130.87685 -12.36972)  Int64Index([1401], dtype='int64')   


In [37]:
poa_bound

gml_id  primaryindex  objectid  poa_code_2016 poa_name_2016  \
0        poa_2016_aust.1             1         1            800          0800   
1        poa_2016_aust.2             2         2            810          0810   
2        poa_2016_aust.3             3         3            812          0812   
3        poa_2016_aust.4             4         4            815          0815   
4        poa_2016_aust.5             5         5            820          0820   
...                  ...           ...       ...            ...           ...   
2665  poa_2016_aust.2639          2639      2639           7268          7268   
2666  poa_2016_aust.2640          2640      2640           7270          7270   
2667  poa_2016_aust.2641          2641      2641           7275          7275   
2668  poa_2016_aust.2642          2642      2642           7276          7276   
2669  poa_2016_aust.2643          2643      2643           7277          7277   

      area_albers_sqkm                                           geometry  \
0               3.1734  POLYGON ((130.83451 -12.45799, 130.83391 -12.4...   
1              23.7902  POLYGON ((130.84711 -12.37749, 130.84731 -12.3...   
2              35.8899  POLYGON ((130.89191 -12.36879, 130.89221 -12.3...   
3               0.6381  POLYGON ((130.87241 -12.37649, 130.87231 -12.3...   
4              39.0462  POLYGON ((130.83501 -12.43009, 130.83511 -12.4...   
...                ...                                                ...   
2665          162.4060  POLYGON ((147.20121 -41.25829, 147.20131 -41.2...   
2666          265.1620  MULTIPOLYGON (((146.82301 -41.13959, 146.82291...   
2667          340.3100  POLYGON ((146.84831 -41.24129, 146.84851 -41.2...   
2668            5.9926  POLYGON ((146.97821 -41.28339, 146.97821 -41.2...   
2669           98.6653  MULTIPOLYGON (((147.08801 -41.38309, 147.08801...   

                         centroid                           sa2_area  
0     POINT (130.84373 -12.46185)  Int64Index([2130], dtype='int64')  
1     POINT (130.87685 -12.36972)  Int64Index([1401], dtype='int64')  
2     POINT (130.91450 -12.37765)  Int64Index([1454], dtype='int64')  
3     POINT (130.86899 -12.37174)  Int64Index([1401], dtype='int64')  
4     POINT (130.86497 -12.42338)  Int64Index([2139], dtype='int64')  
...                           ...                                ...  
2665  POINT (147.21952 -41.29213)  Int64Index([2135], dtype='int64')  
2666  POINT (146.78669 -41.17828)  Int64Index([2135], dtype='int64')  
2667  POINT (146.81144 -41.31036)  Int64Index([2135], dtype='int64')  
2668  POINT (146.96355 -41.28260)                                NaN  
2669  POINT (146.99725 -41.37181)                                NaN  

[2668 rows x 9 columns]

In [46]:
# check the postcodes not allocated
undef_poa_ind = []
for i in range(len(sa2_list)):
    try:
        sa2_list[i] = sa2_list[i][0]
    except IndexError:
        undef_poa_ind.append(i)
        print(i)

38
66
138
206
335
347
349
352
1408
1539
1784
1863
2160
2302
2492
2496
2524
2528
2531
2567
2603
2626


In [52]:
df_undef_poa = (poa_bound[poa_bound.index.isin(undef_poa_ind)]
                [['poa_code_2016', 'poa_name_2016', 'sa2_area']])

In [53]:
df_undef_poa.head()

poa_code_2016 poa_name_2016                       sa2_area
38             880          0880  Int64Index([], dtype='int64')
66            2027          2027  Int64Index([], dtype='int64')
138           2105          2105  Int64Index([], dtype='int64')
206           5575          5575  Int64Index([], dtype='int64')
335           2230          2230  Int64Index([], dtype='int64')

In [70]:
# there are postcodes with multiple SA2s
display(poa_to_sa2.sort_values(by=['POSTCODE', 'PERCENTAGE'], ascending=True).head(20))

POSTCODE  POSTCODE.1  SA2_MAINCODE_2011     SA2_NAME_2011     RATIO  \
1      0800       800.0        701011002.0       Darwin City  1.000000   
13     0810       810.0        701021029.0           Wanguri  0.060207   
5      0810       810.0        701021018.0           Jingili  0.061562   
8      0810       810.0        701021024.0              Moil  0.068575   
6      0810       810.0        701021021.0        Lyons (NT)  0.070201   
12     0810       810.0        701021028.0           Wagaman  0.071421   
2      0810       810.0        701021010.0             Alawa  0.071997   
7      0810       810.0        701021023.0           Millner  0.083483   
11     0810       810.0        701021027.0              Tiwi  0.085041   
3      0810       810.0        701021013.0  Brinkin - Nakara  0.096392   
4      0810       810.0        701021016.0     Coconut Grove  0.096494   
10     0810       810.0        701021026.0       Rapid Creek  0.110723   
9      0810       810.0        701021025.0        Nightcliff  0.123904   
15     0812       812.0        701021014.0     Buffalo Creek  0.000162   
14     0812       812.0        701021011.0             Anula  0.124460   
19     0812       812.0        701021030.0            Wulagi  0.135592   
17     0812       812.0        701021020.0           Leanyer  0.242811   
18     0812       812.0        701021022.0   Malak - Marrara  0.246074   
16     0812       812.0        701021019.0            Karama  0.250887   
29     0820       820.0        701021015.0    Charles Darwin  0.000160   

    PERCENTAGE  
1    99.999998  
13    6.020652  
5     6.156198  
8     6.857527  
6     7.020120  
12    7.142110  
2     7.199707  
7     8.348290  
11    8.504137  
3     9.639178  
4     9.649355  
10   11.072343  
9    12.390383  
15    0.016237  
14   12.446020  
19   13.559184  
17   24.281085  
18   24.607349  
16   25.088725  
29    0.015946

In [72]:
# only get the SA2 with highest percentage
idx = poa_to_sa2.groupby(['POSTCODE'])['PERCENTAGE'].transform(max) == poa_to_sa2['PERCENTAGE']
high_perc_sa2_poa = poa_to_sa2[idx].reset_index(drop=True)

In [73]:
pd.merge(df_undef_poa, high_perc_sa2_poa[['POSTCODE', 'SA2_MAINCODE_2011', 'SA2_NAME_2011']], how='left', left_on='poa_name_2016', right_on='POSTCODE')

poa_code_2016 poa_name_2016                       sa2_area POSTCODE  \
0             880          0880  Int64Index([], dtype='int64')     0880   
1            2027          2027  Int64Index([], dtype='int64')     2027   
2            2105          2105  Int64Index([], dtype='int64')     2105   
3            5575          5575  Int64Index([], dtype='int64')     5575   
4            2230          2230  Int64Index([], dtype='int64')     2230   
5            2261          2261  Int64Index([], dtype='int64')     2261   
6            2263          2263  Int64Index([], dtype='int64')     2263   
7            2267          2267  Int64Index([], dtype='int64')     2267   
8            3960          3960  Int64Index([], dtype='int64')     3960   
9            4178          4178  Int64Index([], dtype='int64')     4178   
10           4802          4802  Int64Index([], dtype='int64')     4802   
11           4875          4875  Int64Index([], dtype='int64')     4875   
12           5606          5606  Int64Index([], dtype='int64')     5606   
13           6168          6168  Int64Index([], dtype='int64')     6168   
14           6530          6530  Int64Index([], dtype='int64')     6530   
15           6537          6537  Int64Index([], dtype='int64')     6537   
16           6707          6707  Int64Index([], dtype='int64')     6707   
17           6713          6713  Int64Index([], dtype='int64')     6713   
18           7001          7001  Int64Index([], dtype='int64')     7001   
19           7001          7001  Int64Index([], dtype='int64')     7001   
20           7150          7150  Int64Index([], dtype='int64')     7150   
21           6799          6799  Int64Index([], dtype='int64')     6799   
22           7175          7175  Int64Index([], dtype='int64')     7175   

    SA2_MAINCODE_2011                  SA2_NAME_2011  
0         702041064.0                      Nhulunbuy  
1         118011343.0     Double Bay - Bellevue Hill  
2         122021421.0      Bayview - Elanora Heights  
3         405041127.0        Yorke Peninsula - North  
4         128011528.0  Cronulla - Kurnell - Bundeena  
5         102021044.0    Bateau Bay - Killarney Vale  
6         102021048.0  Gorokan - Kanwal - Charmhaven  
7         111021220.0        Wangi Wangi - Rathmines  
8         205031087.0                         Foster  
9         301031020.0                         Wynnum  
10        312031359.0           Airlie - Whitsundays  
11        315011401.0                         Torres  
12        406011133.0                   Port Lincoln  
13        507051188.0                     Rockingham  
14        508051210.0                      Geraldton  
15        508021197.0                        Exmouth  
16        508021197.0                        Exmouth  
17        508061223.0                      Roebourne  
18        603011068.0              Wilderness - East  
19        603031073.0             Forestier - Tasman  
20        601041024.0                 Margate - Snug  
21        901021002.0        Cocos (Keeling) Islands  
22        601061034.0        Dodges Ferry - Lewisham

All not allocated POAs were found using the 2011 data.